### DISCUSSION TITLE
Modular/plugins based application possible?

### DISCUSSION THREAD

**Discussion by syaifulnizamyahya** [unoplatform/uno#9147] [Q&A] [answered]
I want to build an app where user can provide their own ui. 
For example, user create a ui page using xaml and save it as plugin. Later my app can load user created plugin which contains ui into my app. The user must follow certain rules though. And must export the ui in plugins forms. 

Is this possible? Any tips? Can i use MEF?

**Comment by jeromelaban** [comment, marked as answer]
Modular applications are definitely possible, but it depends on the platforms. Which are your targets?

**Reply by syaifulnizamyahya** [reply]
My target platform are android, ios, web, and windows.

**Reply by jeromelaban** [reply]
Thanks. iOS does not allow for external binaries to be added to an application (if you want your app to be on the store). The other platforms allow for external code to be added, with the notable difference of WebAssembly. 

For this one, the added code will run interpreted, because there's no JIT to Wasm yet (whether it's for the browsers engine, or for .NET itself..

**Reply by syaifulnizamyahya** [reply]
Ok. External binaries is no go. 
How about xamlloader(just found it)? Im thinking of retreiving the xaml and code behind (as string or binary) from db or web service. And use xamlloader to use it. Is this possible on android, ios, web, and windows?
Thanks.

**Reply by jeromelaban** [reply]
`XamlReader` is available everywhere, it's used by the hot reload engine or by [the playground](https://playground.platform.uno/). While you can load XAML You won't be able to get code-behind C# files with it, though, as it's the same answer as above.
